src/_P103_Atlas_EZO_pH_ORP_EC.ino

```diff
@@ -0,0 +1,641 @@
+            } else if (isFormItemChecked(F("Plugin_103_enable_cal_single"))) {
+                cmd             += PCONFIG(3);
+                triggerCalibrate = true;
+            } else if (isFormItemChecked(F("Plugin_103_enable_cal_L"))) {
+                cmd             += F("low,");
+                cmd             += PCONFIG(4);
+                triggerCalibrate = true;
+            } else if (isFormItemChecked(F("Plugin_103_enable_cal_H"))) {
+                cmd             += F("high,");
+                cmd             += PCONFIG(5);
+                triggerCalibrate = true;
+            }
+            break;
+          }
+      }
+
+      if (triggerCalibrate) {
+        char sensordata[32];
+        _P103_send_I2C_command(I2Cchoice, cmd, sensordata);
+      }
+
+      if(board_type == PH || board_type == EC) {
+        char   deviceTemperatureTemplate[40];
+        String tmpString = web_server.arg(F("Plugin_103_temperature_template"));
+        strncpy(deviceTemperatureTemplate, tmpString.c_str(), sizeof(deviceTemperatureTemplate) - 1);
+        deviceTemperatureTemplate[sizeof(deviceTemperatureTemplate) - 1] = 0; // be sure that our string ends with a \0
+
+        addHtmlError(SaveCustomTaskSettings(event->TaskIndex, (byte *)&deviceTemperatureTemplate, sizeof(deviceTemperatureTemplate)));
+      }
+
+      success = true;
+      break;
+    }
+
+    case PLUGIN_INIT:
+    {
+      break;
+    }
+
+    case PLUGIN_READ:
+    {
+      board_type = PCONFIG(0);
+      I2Cchoice  = PCONFIG(1);
+
+      char sensordata[32];
+      bool status;
+      String readCommand;
+
+      if(board_type == PH || board_type == EC) {
+        // first set the temperature of reading
+        char deviceTemperatureTemplate[40];
+        LoadCustomTaskSettings(event->TaskIndex, (byte *)&deviceTemperatureTemplate, sizeof(deviceTemperatureTemplate));
+
+        String deviceTemperatureTemplateString(deviceTemperatureTemplate);
+        String pooltempString(parseTemplate(deviceTemperatureTemplateString, 40));
+
+        readCommand = F("RT,");
+        double  temperatureReading;
+
+        if (Calculate(pooltempString.c_str(), temperatureReading) != CalculateReturnCode::OK) {
+            temperatureReading = FIXED_TEMP_VALUE;
+        }
+
+        readCommand += temperatureReading;
+
+      } else if(board_type == ORP) {
+        readCommand = F("R,");
+      }
+
+      // ok, now we can read the value
+      status = _P103_send_I2C_command(I2Cchoice, readCommand, sensordata);
+
+
+      // we read the voltagedata char statussensordata[32];
+      char voltagedata[32];
+      status = _P103_send_I2C_command(I2Cchoice, F("Status"), voltagedata);
+
+
+      if (status) {
+        String sensorString(sensordata);
+        String voltage(voltagedata);
+        int    pos = voltage.lastIndexOf(',');
+        UserVar[event->BaseVarIndex]     = sensorString.toFloat();
+        UserVar[event->BaseVarIndex + 1] = voltage.substring(pos + 1).toFloat();
```

Comment:
What if `pos == -1` ?
Please have a look at `validFloatFromString` as that one does also perform checks on expected numerical types.

Reply:
ok